delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/eventlog/EventExchange.scala

```diff
@@ -34,30 +35,57 @@ trait EventExchange {
     *
     * @param event the event from where to compute the state
     */
-  def toState(event: Event, tag: Option[TagLabel] = None): UIO[Option[StateExchange[State]]] =
+  def toState(event: Event, tag: Option[TagLabel] = None): Task[StateExchange[State]] =
     cast.unapply(event) match {
       case Some(e) =>
-        state(e, tag).map(_.map(new StateExchange(_, toExpanded, toSource)))
+        fetch(e, tag).map { s => new StateExchange(s, toExpanded(s.value), toSource(s.value)) }
       case None    =>
-        UIO.pure(None)
+        Task.raiseError(new IllegalArgumentException(""))
```

Comment:
Don't you want to add any error message? Something like 'problem casting an event to an E?' You have the classtag of E so you can even get the runtime name